### DISCUSSION TITLE
SKPaint obsolete members not migrated. [3.116.1]

### DISCUSSION THREAD

**Discussion by dmitryvhf** [mono/SkiaSharp#3104] [General]
I'm updated package from 2.88.3 to 3.116.0/1 version.
Much members of SKPaint class is obsolete now. But not all can migrate because obsolete message incorrect or incomplete. Example,

```
[Obsolete ($"Use {nameof (SKTextAlign)} method overloads instead.")]
public SKTextAlign TextAlign {
   get => SkiaApi.sk_compatpaint_get_text_align (Handle);
   set => SkiaApi.sk_compatpaint_set_text_align (Handle, value);
}
```

What i must use now? Method SKTextAlign not found in sources.

**Comment by mattleibow** [comment]
The text alignment values are no longer part of the paint, but a separate parameter in the DrawText methods. The paint.TextAlign should still work for now, but may be removed at a later date.

But, that obsolete message is also incorrect, it should be `Use the DrawText method overload that accepts a SKTextAlign argument.`

**Reply by dmitryvhf** [reply]
> But, that obsolete message is also incorrect, it should be `Use the DrawText method overload that accepts a SKTextAlign argument.`

Yes. It what i wanted! Thanks for fast response.
P.S. Same problems in few others obsolete messages into SKPaint class.

**Comment by dmitryvhf** [comment]
Problem is not finished. New questions.

SKTextAlign - this property, with Get and Set.
You wrote - "use DrawText method". What you mean? SKCanvas.DrawText? But this action for drawing. Yes, i see new override with SKTextAlign param. This possible good, but i not use this.

What about reading current SKTextAlign state?

**Reply by ToolmakerSteve** [reply]
> What about reading current SKTextAlign state?

For now, you can get it from the `SKPaint` object you pass in to `DrawText`:  
  `canvas.DrawText(text, p, paint.TextAlign, font, paint);`

This will give an "obsolete" warning on `paint.TextAlign`. But it will work.

------------------------------------
If you want to "future proof" your code, in case `paint.TextAlign` goes away some day, then your code needs to hold a separate `TextAlign` object. Wherever is convenient for you.

Would require some refactoring of your code.

How to refactor depends on how you use SKTextAlign today.

Some techniques:
1. Define a static `SKTextAlign` member on one of your classes.
2. Add an `SKTextAlign textAlign` parameter to your methods.

If you decide to refactor, I recommend a "general" technique for storing "state" values, not just `SKTextAlign`:  
  Define a class. Put in it values you want to pass around to different draw methods:

```cs
public class AppDrawState
{
	public static AppDrawState Current = new AppDrawState();

	public SKCanvas Canvas { get; set; }
	public SKFont Font { get; set; }
	public SKTextAlign TextAlign { get; set; }

	// TODO: Other values as desired.

	public void Init(SKCanvas canvas, SKFont font, SKTextAlign textAlign)
	{
		Canvas = canvas;
		Font = font;
		TextAlign = textAlign;
	}
}
```

Usage (1):
```cs
    ...
    AppDrawState.Current.Init(canvas, new SKFont(), SKTextAlign.Left);   // modify as needed.
    ...
    canvas.DrawText("Hello, World", new SKPoint(0, 0), AppDrawState.Current.TextAlign, font, paint);
```

Usage (2):
```cs
public void Method1(SKCanvas canvas)
{
	var state1 = new AppDrawState { Canvas = canvas, TextAlign = SKTextAlign.Center, Font = new SKFont() };
	var paint = new SKPaint { Color = Colors.RoyalBlue.ToSKColor() };
	Method2("Hello, World", paint, state1);
}

public void Method2(string text, SKPaint paint, AppDrawState state)
{
	state.Canvas.DrawText(text, new SKPoint(0, 0), state.TextAlign, state.Font, paint);
}
```

**Reply by dmitryvhf** [reply]
> For now, you can get it from the `SKPaint` object you pass in to `DrawText`: `canvas.DrawText(text, p, paint.TextAlign, font, paint);`
> 
> This will give an "obsolete" warning on `paint.TextAlign`. But it will work.

Yes, sure i looking migration way for next releases.

> If you want to "future proof" your code, in case `paint.TextAlign` goes away some day, then your code needs to hold a separate `TextAlign` object. Wherever is convenient for you.

Huh... Your code interested, but i hope what this will be supported into library.

**Reply by ToolmakerSteve** [reply]
> i hope what this will be supported into library.

Are you asking for SkiaSharp to keep track of the most recently used SKTextAlign value?

So that you could do something like:  
```cs
canvas.DrawText("One", p1, SKTextAlign.Center, font, paint);
canvas.DrawText("Two", p2, SkiaSharp.PreviousTextAlign, font, paint);
```

And "Two" would be displayed with `Center`, because that was the most recent alignment?

**Reply by dmitryvhf** [reply]
> Are you asking for SkiaSharp to keep track of the most recently used SKTextAlign value?

What you mean? Ask SkiaSharp owner or what?

> So that you could do something like:

I dont need DrawText and etc methods. I need readable property (as before).
Example, now i use same like this:

```
                                        if (fontPen.TextAlign != SKTextAlign.Left)
                                        {
                                            float width = result.Width;
                                            if (fontPen.TextAlign == SKTextAlign.Center)
                                            {
                                                width *= 0.5f;
                                            }

                                            x1 -= width;
                                        }
```

I mean - i read TextAlign value for style analize, for next changing.
Where i can use DrawText here? Sure nowhere.

**Comment by Webreaper** [comment]
Yeah, I'm struggling with this too. I've got the code below, but it's completely unclear as to how to migrate it to the new APIs etc. 

It would be helpful if the SkiaSharp samples were updated to replace the obsolete code too, so there are samples we can look at. Any chance you could provide some guidance on how to convert the code below to use the new APIs?

```
            using var font = SKTypeface.FromFamilyName("Arial");
            using var brush = new SKPaint
            {
                Typeface = font,
                TextSize = 64.0f,
                IsAntialias = true,
                Color = new SKColor(255, 255, 255, 255)
            };

            var textWidth = brush.MeasureText(config.WatermarkText);
            var textTargetWidth = targetWidth / 6f;
            var fontScale = textTargetWidth / textWidth;

            brush.TextSize *= fontScale;

            // Offset by text width + 10%
            var rightOffSet = textTargetWidth * 1.1f;

            canvas.DrawText(config.WatermarkText, targetWidth - rightOffSet, targetHeight - brush.TextSize, brush);
```

**Reply by dmitryvhf** [reply]
> Yeah, I'm struggling with this too. I've got the code below, but it's completely unclear as to how to migrate it to the new APIs etc.
> 
> It would be helpful if the SkiaSharp samples were updated to replace the obsolete code too, so there are samples we can look at. Any chance you could provide some guidance on how to convert the code below to use the new APIs?

This no problem to migrate. This your code. You can find changes.

`

            // external values
            float targetWidth = 300;
            float targetHeight = 300;
            string configWatermarkText = "test text";
            SKCanvas canvas = new SKCanvas(null);

            // implementation
            using (SKTypeface typeface = SKTypeface.FromFamilyName("Arial"))
            {
                SKFont font = new SKFont(typeface, 64.0f);
                SKPaint brush = new SKPaint
                {
                    IsAntialias = true,
                    Color = new SKColor(255, 255, 255, 255)
                };

                using (brush)
                {
                    float textWidth = font.MeasureText(configWatermarkText, brush);

                    float textTargetWidth = targetWidth / 6f;
                    float fontScale = textTargetWidth / textWidth;

                    font.Size *= fontScale;

                    // Offset by text width + 10%
                    float rightOffSet = textTargetWidth * 1.1f;

                    SKPoint skPoint = new SKPoint(targetWidth - rightOffSet, targetHeight - font.Size);

                    canvas.DrawText(configWatermarkText, skPoint, font, brush);
                }
            }

`

**Reply by Webreaper** [reply]
Thank you!

**Comment by fox-15** [comment]
Thank you

**Comment by dmitryvhf** [comment]
@mattleibow, sorry, but question is not solved.

Version 3.119.4. SKPaint.TextAlign - obsolete. And obsolete message is not usefull.

How to get TextAlign value? I dont want to using align into DrawText. I want to reading current.

P.S. same for SKPaint.SKTextEncoding
